### PR TITLE
header filter

### PR DESCRIPTION
**Pull request by zhangshen023**
**What this PR does**:

add header filter so that haeders of http request from client can match headers in resource。

**Does this PR introduce a user-facing change?**:
```
name: api name
description: api description
resources:
  - path: '/'
    type: restful
    description: resource documentation
    headers:
      header1: value1 # what this pr does
    filters:
      - filter0
    methods:
      - httpVerb: GET
        onAir: true
        inboundRequest:
          requestType: http
          queryStrings:
            - name: id
              required: false
        integrationRequest:
          requestType: dubbo
          mappingParams:
            - name: queryStrings.id
              mapTo: 1
          applicationName: "BDTService"
          interface: "com.ikurento.user.UserProvider"
          method: "GetUser"
          clusterName: "test_dubbo"
```
```release-note

```

### CODE REVIEW

File: pkg/filter/header/header.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package header
+
+import (
+	"github.com/dubbogo/dubbo-go-proxy/pkg/context"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/context/http"
+	"strings"
```
Comment (cityiron) [inline pkg/filter/header/header.go:23]: split

File: pkg/router/api.go [hunk 2]
```diff
@@ -30,6 +30,7 @@ import (
 type API struct {
 	URLPattern    string `json:"urlPattern" yaml:"urlPattern"`
 	config.Method `json:"method,inline" yaml:"method,inline"`
+	Headers       map[string]string `json:"headers" yaml:"headers"`
```
Comment (cityiron) [inline pkg/router/api.go:33]: add omitempty？

File: pkg/proxy/listener.go [hunk 3]
```diff
@@ -157,6 +158,8 @@ func addFilter(ctx *h.HttpContext, api router.API) {
 		httpFilter(ctx, api.Method.IntegrationRequest)
 	}
 
+	ctx.AppendFilterFunc(header.New().Do())
```
Comment (cityiron) [inline pkg/proxy/listener.go:161]: one rows with RemoteCallFilter

File: pkg/proxy/listener.go [hunk 4]
```diff
@@ -19,6 +19,7 @@ package proxy
 
 import (
 	"context"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/filter/header"
```
Comment (cityiron) [inline pkg/proxy/listener.go:22]: split

**Review by cityiron — CHANGES_REQUESTED** [review]

**Review by yqxu — APPROVED** [review]

**Review by cityiron — APPROVED** [review]

**Review by yqxu — APPROVED** [review]

File: pkg/filter/header/header.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+type headerFilter struct {
+}
+
+// nolint.
+func New() *headerFilter {
+	return &headerFilter{}
+}
+
+func (h *headerFilter) Do() context.FilterFunc {
```
Comment (williamfeng323) [inline pkg/filter/header/header.go:37]: for struct functions that we don't need to modify the fields within the struct,  should we just 
 define a non-pointer function?

**Review by williamfeng323 — APPROVED** [review]

**Pull request merged**